File: ProyectoCivitasRuby/lib/modelo/sorpresa_ir_casilla.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas
  class SorpresaIrCasilla < Sorpresa
    public_class_method :new   
    def initialize(tablero, valor, texto)
      super(texto)
      @tablero = tablero
      @valor = valor
    end
    
    #@Override
    def aplicarajugador(actual, todos)
      if(jugador_correcto(actual, todos))
        informe(actual, todos)
        numero = todos[actual].num_casilla_actual()   
        tirada = @tablero.calcular_tirada(numero, @valor)
        posicion = @tablero.nueva_posicion(numero, tirada)
        todos[actual].mover_a_casilla(posicion)
        @tablero.get_casilla(posicion).recibe_jugador(actual, todos)      
      end
    end
    
    #@Override
    def to_s
      salida = "{@texto}\nSorpresa Ir A Casilla #{@valor}"
    end
    
  end

end

File: ProyectoCivitasJava/src/civitas/SorpresaIrCasilla.java
/**
 * Proyecto Civitas
 * Sergio Hervás Cobo
 * Javier Suárez Lloréns
 * Programación y Diseño Orientado a Objetos
 * 2º B3 - Grado en Ingeniería Informática
 * E.T.S.I. Informática y de Telecomunicación
 * Universidad de Granada
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author sergio
 */
public class SorpresaIrCasilla extends Sorpresa {

    private int valor;
    private Tablero tablero;

    SorpresaIrCasilla(Tablero tablero, int valor, String texto) {
        super(texto);
        this.tablero = tablero;
        this.valor = valor;
    }

    @Override
    void aplicarAJugador(int actual, ArrayList<Jugador> todos) {
        int casillaActual, tirada, posicion;
        if (jugadorCorrecto(actual, todos)) {
            super.informe(actual, todos);
            casillaActual = todos.get(actual).getNumCasillaActual();
            tirada = tablero.calcularTirada(casillaActual, valor);
            posicion = tablero.nuevaPosicion(casillaActual, tirada);
            todos.get(actual).moverACasilla(posicion);
            tablero.getCasilla(posicion).recibeJugador(actual, todos);
        }
    }
   
    @Override
    public String toString(){
        String salida = "Sorpresa Ir A Casilla " + valor;
        return salida;
    }
}
